keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,146 @@
+
+## Motivation
+
+Currently only string parameters are supported by KUDO. This limits the possible use-cases of parameters. Various use cases benefit from having parameters that are lists or dictionaries. While some of this functionality can be implemented in Go templates using [Sprig functions][1], this is unintuitive.
+Having parameter types creates the possibility for parameter validation.
+
+### Goals
+
+- Provide operator developers with more versatile parameters
+- Allow the definition of validation functions for parameters
+
+### Non-Goals
+
+- Change how parameters are defined
+
+## Proposal
+
+### Support Optional Parameter Types
+
+By adding an optional `type` field to the list of parameters, parameter types can be specified. If this field isn't present, it defaults to the `string` type. Other types like `list`, `dict` or `integer` are possible. KUDO's renderer uses the type information to convert the string value to the respective type. For example, for the `dict` type, the input value would be YAML that is unmarshalled to a `map[string]interface{}`. Go's template engine allows to access and iterate list and dictionary items.
+
+Consider a cluster that has nodes in multiple regions. We want to ensure that a nginx deployment has 2 replicas in one region and 1 replica in the other one. Furthermore they should support both HTTP as well as HTTPS, hence need a list of port definitions. These parameters are described by `list` and `dict` types:
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default: |
+      - region: us-east-1:
+        replicas: 2
+      - region: us-west-1:
+        replicas: 1
+```
+
+The deployment template then iterates over the items of the respective parameters:
+
+```yaml
+{{ $ports := .Params.ports }}
+{{ range .Params.topology }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app.kubernetes.io/name: nginx
+spec:
+  replicas: {{ .replicas }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: nginx
+  template:
+    metadata:
+      name: nginx
+      labels:
+        app.kubernetes.io/name: nginx
+    spec:
+      nodeSelector:
+        topology.kubernetes.io/region: {{ .region }}
+      containers:
+      - name: nginx
+        image: nginx:stable
+        ports:
+        {{ range $ports }}
+        - containerPort: {{ . }}
+        {{ end }}
+{{ end }}
+```
+
+### Provide Parameter Validation
+
+Adding types for lists and dictionaries provides access to the items of these structures in templates. Adding types like integers won't provide benefits for templating, but would allow for input value validation. KUDO could, e.g., validate that an integer isn't too large, that a list doesn't have too many entries, or that a string isn't too long. Adding parameter validation is independent from adding parameter types.
```

Comment:
Another thing useful to a user would be to verify that more complex objects passed as parameters have correct structure, even before it's merged with the rest of the (potentially huge) template (at which point it might be a little overwhelming to figure out what an error message means).